README.md
Python version of the Platefit continuum fitting.

By Madusha Gunawardhana (gunawardhana@strw.leidenuniv.nl)


docs/changelog.rst
Changelog
=========
v0.7 (09/06/2021)
-----------------

New features
^^^^^^^^^^^^
- new MCMC error estimate (see tutorial)
- change default gamma init value to 2
- updated tutorial

Breaking Changes
^^^^^^^^^^^^^^^^

Bug fixes
^^^^^^^^^
- fix two bugs with masked data
- update SNR estimation for blended lines, now use sqrt(snr**2)

v0.6 (30/01/2021)
-----------------
New features
^^^^^^^^^^^^
- allow to use any minimizing methods and its corresponding input parameters from scipy minimize
- set default method to least_square using Trust Region Reflective method with 1.e-3 tolerance on parameters
- save minimization method and status in ztable
- add infoline and infoz option in plot_fit to display values from the tables lines and ztable
- add line multiplet (eg OII3727b = OII3726 + OII3729) in the returned line table.
- update default line ratio range of OII doublet to 0.3-1.5
- one can now specify his own LSF model
- updated documentation, now based on a notebook tutorial

Breaking Changes
^^^^^^^^^^^^^^^^
- remove bootstrap option which was not given robust results and is now better replace by Trust Region Reflective least_squares 
- name of emission lines in default table have been updated to reflect common usage
- updated emission lines 

Bug fixes
^^^^^^^^^
- it is now possible to fix values by setting min=init=max in linepars dict (eg vdisp=(50,50,50).
- fix a bug with the option flag major_lines not correctly propagated


v0.5 (17/01/2021)
-----------------
New features
^^^^^^^^^^^^
- add option n_cpu to speed up bootstrap computation using parallel computing
- add option fitlines to perform abs fitting only (Leindert)
- complete the reference line table with some UV emission lines
- add absorption line fitting
- compute line fit quality estimate for individual lines 
- replace the emcee method with bootstrap
- replace the mpdaf linelist module by a specific one in pyplatefit

Breaking Changes
^^^^^^^^^^^^^^^^
- Option comp_bic has been supressed in fit_spec

Bug fixes
^^^^^^^^^
- correct a bug in redshift value and error reported in ztable
- correct a bug in redshift error estimation


v0.4 (24/03/2020)
-----------------
New features
^^^^^^^^^^^^
- add the fit of double lyman-alpha line
- allow the use of its own lines table
- allow to define separately the bounds and starting value for lya
- update documentation 
- add the name of the line with maximum SNR in ztable
- line_plot function 
- the bounds for EMCEE fit are reduced and set with respect to the LSQ first fit
- improve unit tests 


Bug fixes
^^^^^^^^^
- correct bugs when using AO spectra with masked Na wavelength range 
- correct a bug in BIC computation for lyalpha
- set correctly the initial value of velocity dispersion for the fit
- correct a bug in the estimation of LBDA_OBS for lines other than lyalpha
- filter emcee warnings


v0.3 (19/10/2019)
--------------------
New features
^^^^^^^^^^^^
- Complete reorganisation and reoptimisation of the code
- Add documentation and tutorial


v0.2 (19/04/2019)
-----------------

Full rewriting of line fitting


v0.1 (12/04/2019)
-----------------

First released version


muse_spectrum.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import (division, print_function, absolute_import,
                        unicode_literals)
# __all__ = ["format_output"]

import sys

import numpy as np
from mpdaf.sdetect import Source
from mpdaf.obj.spectrum import Spectrum
from astropy.table import Table

import matplotlib.pyplot as plt
from pyplatefit import __version__
from pyplatefit.platefit import Platefit, fit_spec, plot_fit

import logging

logger = logging.getLogger('pyplatefit')
logger.setLevel(logging.DEBUG)

logger.info('pyplatefit version %s', __version__)
debug = True

vdisp = 80.0

name = '/Users/rolandbacon/Dropbox/Soft/python/pyplatefit/tests/test_data/udf10_00605.fits'
zinit = 3.43774

#name = '/Users/rolandbacon/Dropbox/Soft/python/pyplatefit/tests/test_data/udf10_00723.fits'
#z = 3.18817

#name = '/Users/rolandbacon/Dropbox/Soft/python/pyplatefit/tests/test_data/udf10_00056.fits'
#z = 1.30604

#name = '/Users/rolandbacon/Dropbox/MUSE/GTO/UDF/DR2/UDF10/Final/DR2_udf10_000043.fits'
#z = 1.09989

#name = '/Users/rolandbacon/Dropbox/MUSE/GTO/UDF/DR2/UDF10/Final/tests/sptest.fits'
#zinit = 3.17140
#z = 3.17136

#z = z + 400/300000.0
#print(z)

#name = '/Users/rolandbacon/Dropbox/MUSE/Anna/spec_LAE_stack_platefit.fits'
#zinit = 0

name = '/Users/rolandbacon/Dropbox/Soft/python/pyplatefit/tests/test_data/udf10_00106.fits'
zinit = 3.27554

sp = Spectrum(name)
res = fit_spec(sp, zinit, lines=['LYALPHA'], emcee=False, dble_lyafit=True, find_lya_vel_offset=False)

#res2 = fit_spec(sp, z)
#                linepars=dict(delta_vel=100, delta_vdisp=50, delta_gamma=None))

fig,ax = plt.subplots(1,1,figsize=(10,5))
plot_fit(ax, res, line='LYALPHA1', line_only=True)
res['ztable'].pprint_all()
print(res['lines'].loc['LYALPHA1'])
print(res['lines'].loc['LYALPHA2'])

#ax[0].set_title('LSQ Fit')
#plot_fit(ax[1], res2, line='CIV1548', line_only=True, start=True)
#ax[1].set_title('EMCEE Fit')
#plot_fit(ax, res, filterspec=20)
plt.show()

#res = fit_spec(sp, z, emcee=True, ziter=False, find_lya_vel_offset=True, use_line_ratios=True)
#res = fit_spec(sp, z, emcee=True)
#res = fit_spec(sp, z, emcee=True, linepars=dict(progress=True))




#pl = Platefit()
#res = pl.fit(sp, z)
#pl.fit(sp, z, emcee=False, lines=['LYALPHA'])


#pl.info(res)
#fig,ax = plt.subplots(1,1)
##pl.plot_cont(ax, res)
#pl.plot_lines(ax, res, line=)
#plt.show()

#res = fit_spec(sp, z, emcee=False, comp_bic=True, lines=['OII3727','OII3729'], use_line_ratios=True,
               #linepars=dict(line_ratios=[("OII3727", "OII3729", 0.5, 0.8)])
               #)
#res = fit_spec(sp, z, ziter=False, lines=['OII3727','OII3729'], use_line_ratios=True, 
               #linepars=dict(line_ratios=[("OII3727", "OII3729", 0.5, 0.8)]))


#pl = Platefit(linepars=dict(steps=100))
#res2 = pl.fit(sp, z, emcee=False, vel_uniq_offset=False, eqw=True, trimm_spec=True)
#pl.info(res2)

#if 'OII3727' in res['linetable']['LINE']:
    #lines = ['OII3727','OII3729']
#res2 = pl.fit_lines(res['line'], z, lines=lines, trimm_spec=True)
#pl.info_lines(res2, full_output=False)
#res3 = pl.fit_lines(res['line'], z, lines=lines, trimm_spec=False)
#pl.info_lines(res3, full_output=False)


#fig,ax = plt.subplots(1,1)
#pl.plot(ax, res, line='HALPHA', margin=30)
##pl.plot_lines(ax, res['res_line'], line='HALPHA', margin=30)
#plt.show()




#res_cont = pl.fit_cont(sp, z, vdisp)
#pl.info_cont(res_cont)
#res_line = pl.fit_lines(res_cont['line_spec'], z, lines=['OII3727','OII3729'])
#res_line = pl.fit_lines(res_cont['line_spec'], z, emcee=False, major_lines=True, use_line_ratios=True)
#pl.info_lines(res_line)
#print(res_line.linetable)


#fig,ax = plt.subplots(1,2)
#pl.plot(ax, {**res_line,**res_cont})
#plt.show()

#pl.plot(ax, res_line)
#plt.show()




print('end')
